File: leetcode/completed/767.cpp
#include <string>

using namespace std;

class Solution {
public:
    string reorganizeString(string s) {
        int counter[26] = {0};
        char max_char;
        int max_freq = 0;
        int n = s.size();

        for (auto& c : s) {
            counter[c - 'a']++;
            if (counter[c - 'a'] > max_freq) {
                max_char = c;
                max_freq = counter[c - 'a'];
            }
        }

        if (max_freq > (n + 1)/2) return "";

        int i = 0;

        while(max_freq-- > 0) {
            s[i] = max_char;
            i += 2;
        }
        counter[max_char - 'a'] = 0;

        for (int j = 0; j < 26; j++) {
            while (counter[j]-- > 0) {
                if (i >= n) i = 1;
                s[i] = 'a' + j;
                i += 2;
            }
        }

        return s;

    }
};

File: leetcode/completed/767.py
from heapq import heappop, heappush, heapify

class Solution:
    def reorganizeString(self, s: str) -> str:
        counter = {}
        for c in s:
            counter[c] = counter.get(c, 0) + 1
        
        heap = [(-counter[key], key) for key in counter.keys()]
        heapify(heap)

        res = ''
        while heap:
            first_cnt, first_chr = heappop(heap)
            if not res or res[-1] != first_chr:
                res += first_chr
                if first_cnt + 1 != 0:
                    heappush(heap, (first_cnt + 1, first_chr))
            else:
                if not heap: return ''
                second_cnt, second_chr = heappop(heap)
                res += second_chr
                if second_cnt + 1 != 0:
                    heappush(heap, (second_cnt + 1, second_chr))
                heappush(heap, (first_cnt, first_chr))

        return res
    
class Solution:
    def reorganizeString(self, s: str) -> str:
        counter = {}
        for c in s:
            counter[c] = counter.get(c, 0) + 1
        
        heap = [(-counter[key], key) for key in counter.keys()]
        heapify(heap)

        res = ''
        while len(heap) > 1:
            first_cnt, first_chr = heappop(heap)
            second_cnt, second_chr = heappop(heap)
            res += first_chr
            res += second_chr
            if first_cnt + 1 != 0:
                heappush(heap, (first_cnt + 1, first_chr))
            if second_cnt + 1 != 0:
                heappush(heap, (second_cnt + 1, second_chr))

        if heap:
            if heap[0][0] == -1:
                res += heap[0][1]
            else:
                res = ''

        return res
    
class Solution:
    def reorganizeString(self, s: str) -> str:
        counter = {}
        max_key = ''
        max_freq = 0
        n = len(s)

        for c in s:
            counter[c] = counter.get(c, 0) + 1
            if counter[c] > max_freq:
                max_freq = counter[c]
                max_key = c

        if max_freq > (len(s)+1) // 2:
            return ''

        res = ['']*n
        i = 0

        for _ in range(counter[max_key]):

            res[i] = max_key
            i += 2

        del counter[max_key]

        for key in counter:
            for _ in range(counter[key]):
                
                if i >= n: i = 1

                res[i] = key
                i += 2

        return ''.join(res)